### PR TITLE
RFD 115 - Teleport Oracle DB Access Integration

### PR DESCRIPTION
**Pull request by smallinsky**
### What: 
RFD 115 - Teleport Oracle DB Access Integration

### CODE REVIEW

File: rfd/0084-oracle-db-access-integration.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+
+This RFD proposes a way to integrate Oracle Database Access with Teleport.
+Oracle DB Access differs from currently supported database protocols by Teleport Database Access.
+Namely, Oracle DB protocol is private and there is not any official documentation describing the Oracle protocol.
+
+## Why
+
+We want to increase the number of databases supported by Teleport and allow users to connect to Oracle using Teleport Database Access.
+Where we want to provide a way for users to connect to Oracle databases through Teleport using the same Teleport UI/UX as for other supported databases.
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:18]: I would rephrase this a little to highlight mutual benefits e.g. something like:

```
Adding Oracle support to Teleport Database Access would allow our shared customer bases to take advantage of Teleport's access controls and audit logging capabilities when accessing their Oracle databases. 

* Administrators will be able to control access to their entire fleet of database servers through their identity provider with SSO.
* Users will be able to connect to Oracle databases with the tools they're familiar with without having to deal with passwords or shared secrets.
* Auditors will be able to view the database activity and tie it to a particular user identity within an organization.
```

File: rfd/0084-oracle-db-access-integration.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+
+# Scope of Integration
+- **Teleport as Oracle Access Proxy**: Teleport Oracle DB agents should be able to act like a proxy between the incoming Oracle client connection and connection to Oracle Server where the Teleport DB Agent will terminate the incoming TLS connection and establish a new TLS connection to the Oracle Server using a new TLS Certificate and forward the traffic between Oracle client and server.
+- **Audit Logging** (Optional): After TLS termination of incoming client Oracle connection Teleport DB Agents needs to be able to parse the Oracle wire protocol to provide Teleport audit logs and audit client interaction with Oracle database.
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:23]:
```suggestion
- **Audit Logging**: Teleport needs to be able to inspect the Oracle wire protocol to provide Teleport audit logs and audit client interaction with Oracle database.
```

File: rfd/0084-oracle-db-access-integration.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+Namely, Oracle DB protocol is private and there is not any official documentation describing the Oracle protocol.
+
+## Why
+
+We want to increase the number of databases supported by Teleport and allow users to connect to Oracle using Teleport Database Access.
+Where we want to provide a way for users to connect to Oracle databases through Teleport using the same Teleport UI/UX as for other supported databases.
+
+
+# Scope of Integration
+- **Teleport as Oracle Access Proxy**: Teleport Oracle DB agents should be able to act like a proxy between the incoming Oracle client connection and connection to Oracle Server where the Teleport DB Agent will terminate the incoming TLS connection and establish a new TLS connection to the Oracle Server using a new TLS Certificate and forward the traffic between Oracle client and server.
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:22]:
```suggestion
- **Teleport as Oracle Access Proxy**: Teleport should be able to act like a proxy between the incoming Oracle client connection and connection to Oracle Server where the Teleport will terminate the incoming TLS connection and establish a new TLS connection to the Oracle Server using a new TLS Certificate and forward the traffic between Oracle client and server.
```

File: rfd/0084-oracle-db-access-integration.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+- **Audit Logging** (Optional): After TLS termination of incoming client Oracle connection Teleport DB Agents needs to be able to parse the Oracle wire protocol to provide Teleport audit logs and audit client interaction with Oracle database.
+
+
+### TLS Termination of Incoming connection:
+Teleport needs to be able TLS terminate incoming Oracle client connections to replace the TLS certificate with a new one signed by Teleport Database CA and establish a new TLS connection to the Oracle Server.
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:27]:
```suggestion
Teleport needs to be able to TLS-terminate incoming Oracle client connection and reestablish a new TLS connection that uses Teleport-signed client certificate to the Oracle Server. Oracle database server will need to be configured to trust Teleport's certificate authority to be able to validate client connections.
```

File: rfd/0084-oracle-db-access-integration.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+It seems that TLS connection between Oracle Client and Oracle Server can be renegotiated by some mechanism in L7 Oracle protocol. To support TLS termination on Teleport proxy side the TLS renegotiation triggered by Oracle Server needs to be also handled on Teleport DB Agent side.
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:28]: Let's remove this.
```suggestion
```

File: rfd/0084-oracle-db-access-integration.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+#### Oracle Client:
+
+Oracle clients support TLS connections to the Oracle Server by using a custom container called [Oracle Wallet](https://docs.oracle.com/cd/E92519_02/pt856pbr3/eng/pt/tsvt/concept_UnderstandingOracleWallet.html#:~:text=Oracle%20Wallet%20is%20a%20container,is%20used%20for%20security%20credentials.
+) that stores authentication credentials and certificates.
+Teleport `tsh db login` command for Oracle database will generate cert in Oracle Wallet format allowing to configure the wallet in Oracle database clients like [sqlplus](https://docs.oracle.com/cd/B19306_01/server.102/b14357/qstart.htm) or [SQL Oracle Developer](https://www.oracle.com/database/sqldeveloper/)
+
+
+##### UX:
+
+Oracle will integrate with teleport in the same way as other databases.
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:44]:
```suggestion
Teleport will integrate with Oracle in the same way as other databases.
```

File: rfd/0084-oracle-db-access-integration.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+* `tsh db connect` - would start [sqlplus](https://docs.oracle.com/cd/B19306_01/server.102/b14357/qstart.htm) Oracle CLI.
+* `tsh proxy db` - would start proxy for 3rd party GUI clients like  [SQL Oracle Developer](https://www.oracle.com/database/sqldeveloper/)
+
+
+
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:50]:
```suggestion
```

File: rfd/0084-oracle-db-access-integration.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+#### Oracle Server Setup:
+The new `tctl auth sign` `--format=oracle` sign format will be introduced where Teleport Database CA authority and generated certificate/key pair will be store in Oracle Wallet SSO autologin format:
```
Comment (r0mant) [inline rfd/0084-oracle-db-access-integration.md:53]:
```suggestion
The new `tctl auth sign` `--format=oracle` sign format will be introduced where Teleport certificate authority and generated certificate/key pair will be stored in Oracle Wallet SSO autologin format:
```

File: rfd/0084-oracle-db-access-integration.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+Teleport needs to be able to TLS-terminate incoming Oracle client connection and reestablish a new TLS connection that uses Teleport-signed client certificate to the Oracle Server. Oracle database server will need to be configured to trust Teleport's certificate authority to be able to validate client connections.
+
+
+## Details
+
+### Teleport Database Access Configuration:
+
+#### Oracle Client:
+
+Oracle clients support TLS connections to the Oracle Server by using a custom container called [Oracle Wallet](https://docs.oracle.com/cd/E92519_02/pt856pbr3/eng/pt/tsvt/concept_UnderstandingOracleWallet.html#:~:text=Oracle%20Wallet%20is%20a%20container,is%20used%20for%20security%20credentials.
```
Comment (stevenGravy) [inline rfd/0084-oracle-db-access-integration.md:39]: Do we have a specific set of Oracle versions we will support, ex: 15+?

Comment (smallinsky) [reply rfd/0084-oracle-db-access-integration.md:39]: Yes, I'm in the process of testing the integration with other Oracle versions. I will updated the RFD soon.

Comment (smallinsky) [reply rfd/0084-oracle-db-access-integration.md]: @stevenGravy 
I have added a section about Oracle server requirement: https://github.com/gravitational/teleport/pull/16435/files#diff-a348bce1705dbd09c2ec0ab96d7e4fa49b7f6f5b4413cd9e07fe0986e1bd904dR40

File: rfd/0112-oracle-db-access-integration.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+authors: Marek Smoliński (marek@goteleport.com)
+state: draft
+---
+## Required Approvers
+
+- Engineering: `@r0mant`
+- Product: `@klizhentas || @xinding33`
+- Security: `@reedloden || @jentfoo`
+
+# RFD 84 - Teleport Oracle Access Integration
```
Comment (r0mant) [inline rfd/0112-oracle-db-access-integration.md:12]: Nit: Update RFD number.

File: rfd/0112-oracle-db-access-integration.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+## What
+
+
+This RFD outlines the scope of Oracle integration with Teleport Database Access.
+Teleport Database Oracle Access integration with leverage the details about Oracle Protocol shared in Official Oracle GH repository  https://github.com/oracle/python-oracledb/tree/main/src/oracledb/impl released under Apache License, Version 2.0.
```
Comment (r0mant) [inline rfd/0112-oracle-db-access-integration.md:18]:
```suggestion
Teleport Database Oracle Access integration will leverage the details about Oracle Protocol shared in Official Oracle GH repository  https://github.com/oracle/python-oracledb/tree/main/src/oracledb/impl released under Apache License, Version 2.0.
```

File: rfd/0112-oracle-db-access-integration.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+\
+[SQLNET.AUTHENTICATION_SERVICES](https://docs.oracle.com/cd/E11882_01/network.112/e10835/sqlnet.htm#NETRF2035)
+\
+[SSL_CLIENT_AUTHENTICATION](https://docs.oracle.com/cd/E11882_01/network.112/e10835/sqlnet.htm#NETRF233)
+
+#### Create a OracleDB User wth TLS x509 DN Authentication:
+Oracle server allows to authenticate database user based on the certificate CN field:
+
+```azure
+CREATE USER alice IDENTIFIED EXTERNALLY AS 'CN=alice;
```
Comment (r0mant) [inline rfd/0112-oracle-db-access-integration.md:201]:
```suggestion
CREATE USER alice IDENTIFIED EXTERNALLY AS 'CN=alice';
```

**Review by r0mant — APPROVED** [review]

File: rfd/0084-oracle-db-access-integration.md (hunk 9 above)
Comment (stevenGravy) [reply rfd/0084-oracle-db-access-integration.md]: thanks!

**Review by greedy52 — APPROVED** [review]

**Comment by r0mant** [comment]
@smallinsky Can this be merged?
